examples/MLIRVector/vector-broadcast.mlir

```diff
@@ -17,6 +19,117 @@ func.func @main() -> i32 {
   %broadcast_vec_2d = vector.broadcast %load_vec : vector<4xf32> to vector<4x4xf32>
   vector.print %broadcast_vec_2d : vector<4x4xf32>
 
+  // Detailed example for all cases in broadcast:
+  %v1 = arith.constant dense<[[1.0], [2.0], [3.0]]> : vector<3x1xf32>
+  %v2 = arith.constant dense<[[[1.0], [2.0], [3.0]]]> : vector<1x3x1xf32>
+
+  func.call @broadcast_scalar_to_vector(%f1) : (f32) -> vector<3x1xf32>
+  func.call @broadcast_low_dim_to_high_dim(%v1) : (vector<3x1xf32>) -> vector<4x3x2xf32>
+  func.call @broadcast_1_to_n_case(%v2) : (vector<1x3x1xf32>) -> vector<4x3x2xf32>
+  func.call @broadcast_n_to_n_case(%v1) : (vector<3x1xf32>) -> vector<3x2xf32>
+
   %ret = arith.constant 0 : i32
   return %ret : i32
 }
+
+func.func @broadcast_scalar_to_vector(%src: f32) -> vector<3x1xf32> {
+  // broadcast a scalar to vector
+  %result = vector.broadcast %src : f32 to vector<3x1xf32>
+
+  // equals to splat it, or, "fill" it
+  // like [[%src], [%src], [%src]]
+  %w = vector.splat %src : vector<3x1xf32>
+
+  // %w == %result
+  vector.print %result : vector<3x1xf32>
+  vector.print %w : vector<3x1xf32>
+
+  return %result : vector<3x1xf32>
+}
+
+func.func @broadcast_low_dim_to_high_dim(%src: vector<3x1xf32>) -> vector<4x3x2xf32> {
+  // broadcast vector from low dimension to high can be transform into 
+  // broadcasting two vectors with same dimension
+  %result = vector.broadcast %src : vector<3x1xf32> to vector<4x3x2xf32>
+
+  // vector<... x T> <=> vector<1 x ... x T> <=> vector<1 x 1 x ... x T>
+  // <=> vector<1 x 1 x ... x 1 x ... x T>
+  // it is obviously that we can extend a vector to any higher dimension by
+  // adding dimensions with length "1" in front of it, like this:
+  %zero = arith.constant 0.0 : f32
+  %t0 = vector.broadcast %zero : f32 to vector<1x3x1xf32>
+  %t1 = vector.insert %src, %t0[0] : vector<3x1xf32> into vector<1x3x1xf32>
+
+  // then we have two vectors with same dimension, which can be broadcasted by:
+  %t2 = func.call @broadcast_1_to_n_case(%t1) : (vector<1x3x1xf32>) -> vector<4x3x2xf32>
+
+  // %result == %t2
+  vector.print %result : vector<4x3x2xf32>
+  vector.print %t2 : vector<4x3x2xf32>
+
+  return %result : vector<4x3x2xf32>
+}
+
+func.func @broadcast_1_to_n_case(%src: vector<1x3x1xf32>) -> vector<4x3x2xf32> {
+  // Case 1, "1->n" kind. 
+  //    broadcast %src: vector<1 x ... x T> to vector<n x ... x T> 
```

Comment:
Please remove the unnecessary whitespace at the start of this comment. Take care of this at all relevant places in the PR.

Reply:
I use these indentions as "code block" in Markdown for pseudo codes and variables that I want to emphasize. Is this an unrecommended usage?